backend/apis/leads/add-leads.api.js

```diff
@@ -0,0 +1,116 @@
+const mysql = require("mysql");
+const { app, databaseError, pool, invalidRequest } = require("../../server");
+const {
+  checkValid,
+  nonEmptyString,
+  validArray,
+  validBoolean,
+  validDate,
+  validInteger
+} = require("../utils/validation-utils");
+
+app.post("/api/leads", (req, res) => {
+  const leads = req.body.data;
+  const user = req.session.passport.user;
+
+  if (!Array.isArray(leads)) {
+    return invalidRequest(res, [
+      `POST /api/leads must be called with an array of leads in the request body`
+    ]);
+  }
+
+  for (let i = 0; i < leads.length; i++) {
+    const validityErrors = checkValid(
+      leads[i],
+      validDate("dateOfSignUp"),
+      nonEmptyString("firstName"),
+      nonEmptyString("lastName"),
+      nonEmptyString("phone"),
+      validBoolean("smsConsent"),
+      nonEmptyString("zip"),
+      validInteger("age"),
+      nonEmptyString("gender"),
+      validArray("eventSources", validInteger),
+      validArray("leadServices", validInteger)
+    );
+
+    if (validityErrors.length > 0) {
+      return invalidRequest(res, validityErrors);
+    }
+  }
+
+  let leadQuery = "";
+  const leadDataArray = [];
+
+  for (let i = 0; i < leads.length; i++) {
+    const lead = leads[i];
+
+    const addToLeadQuery =
+      "INSERT INTO leads (dateOfSignUp, firstName, lastName, phone, smsConsent, zip, age, gender, addedBy, modifiedBy) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)";
+    const setLeadId = "SET @leadId = LAST_INSERT_ID()";
+
+    leadDataArray.push(
+      lead.dateOfSignUp,
+      lead.firstName,
+      lead.lastName,
+      lead.phone,
+      lead.smsConsent,
+      lead.zip,
+      lead.age,
+      lead.gender,
+      user.id,
+      user.id
+    );
+
+    const leadServices = lead.leadServices;
+    let leadServicesQuery = "";
+
+    if (leadServices.length > 0) {
+      leadServicesQuery =
+        "INSERT INTO leadServices (leadId, serviceId) VALUES (@leadId, ?)";
+
+      leadDataArray.push(leadServices[0]);
+
+      if (leadServices.length > 1) {
+        for (let j = 1; j < leadServices.length; j++) {
+          const serviceId = leadServices[j];
+          leadServicesQuery = leadServicesQuery + ", (@leadId, ?)";
+          leadDataArray.push(serviceId);
+        }
+      }
+    }
+
+    const leadEvents = lead.eventSources;
+    let leadEventsQuery = "";
+
+    if (leadEvents > 0) {
+      leadEventsQuery =
+        "INSERT INTO leadEvents (leadId, eventId) VALUES (@leadId, ?)";
+
+      leadDataArray.push(leadEvents[0]);
+
+      if (leadEvents.length > 1) {
+        for (let j = 1; j < leadEvents.length; j++) {
+          const eventId = leadEvents[j];
+          leadEventsQuery = leadEventsQuery + ", (@leadId, ?)";
+          leadDataArray.push(eventId);
+        }
+      }
+    }
+
+    const newLeadQuery = `${addToLeadQuery}; ${setLeadId}; ${leadServicesQuery}; ${leadEventsQuery};`;
+    leadQuery = `${leadQuery} ${newLeadQuery}`;
```

Comment:
From what I can tell, `leadQuery` always is an empty string at this point. Do we even need to have both `newLeadQuery` and `leadQuery` variables? Would something like this work?

```js
leadQuery = `${addToLeadQuery}; ${setLeadId}; ${leadServicesQuery}; ${leadEventsQuery};`
```

Reply:
Wouldn't that only make the query for a single lead though? leadQuery is the entire query together for all the leads and newLeadQuery is the one for the single lead built within the loop and then added to leadQuery at the end. Should I be doing:

`leadQuery = `${leadQuery} ${addToLeadQuery}; ${setLeadId}; ${leadServicesQuery}; ${leadEventsQuery};``

Reply:
Ah I see! Yeah I missed that this was inside of the for loop. Disregard my previous comment 👍